### PR TITLE
made serializer typings simpler

### PR DESCRIPTION
**Pull request by Goldziher**
PR introducing the changes suggested by @kalekseev in #97

### CODE REVIEW

File: rest_framework-stubs/serializers.pyi [hunk 1]
```diff
@@ -213,21 +204,22 @@ class ListSerializer(
...
+    def validate(self, attrs: Any) -> Any: ...
     @property
     def data(self) -> ReturnList: ...
     @property
     def errors(self) -> ReturnList: ...
 
 def raise_errors_on_nested_writes(method_name: str, serializer: BaseSerializer, validated_data: Any) -> NoReturn: ...
 
-class ModelSerializer(Serializer, BaseSerializer[_MT, Mapping[str, Any], Dict[str, Any], _MT]):
+class ModelSerializer(Serializer, BaseSerializer[_MT]):
```
Comment (sobolevn) [inline rest_framework-stubs/serializers.pyi:215]: Let's add a bunch of tests for this type! It would be super helpful in the future.

**Comment by sobolevn** [comment]
@kalekseev I would love to hear your thoughts.

File: rest_framework-stubs/serializers.pyi [hunk 2]
```diff
@@ -144,19 +141,16 @@ def as_serializer_error(exc: Exception) -> Dict[str, List[ErrorDetail]]: ...
 class Serializer(
     BaseSerializer[
         Union[_MT, Mapping[str, Any]],
```
Comment (kalekseev) [inline rest_framework-stubs/serializers.pyi:143]:
```suggestion
        Union[_IN, Mapping[str, Any]],
```
_MT is bound to Model so it should be used only with ModelSerializer

File: rest_framework-stubs/serializers.pyi [hunk 3]
```diff
@@ -87,31 +87,28 @@ ALL_FIELDS: str = ...
 
 _MT = TypeVar("_MT", bound=Model)  # Model Type
 _IN = TypeVar("_IN")  # Instance Type
-_DT = TypeVar("_DT")  # Data Type
-_VT = TypeVar("_VT")  # Value Type
-_RP = TypeVar("_RP")  # Representation Type
 
-class BaseSerializer(Generic[_VT, _DT, _RP, _IN], Field[_VT, _DT, _RP, _IN]):
+class BaseSerializer(Generic[_IN], Field[Any, Any, Any, _IN]):
```
Comment (kalekseev) [inline rest_framework-stubs/serializers.pyi:91]: If we are going with all these generic params in Field I would move _IN to first place to make it uniform with Serializer

File: rest_framework-stubs/serializers.pyi (hunk 2 above)
Comment (kalekseev) [reply rest_framework-stubs/serializers.pyi:143]: No need to make this generic this specific, let's use just _IN, someone can create subclass that will work with arbitrary objects not models or dict

File: rest_framework-stubs/serializers.pyi [hunk 4]
```diff
@@ -171,9 +165,6 @@ class Serializer(
 
 class ListSerializer(
     BaseSerializer[
-        List[Any],
-        Union[MultiValueDict, List[Any]],
-        List[Dict[str, Any]],
         List[Dict[str, Any]],
```
Comment (kalekseev) [inline rest_framework-stubs/serializers.pyi:168]: same problem here, no need to make such constraints on user, for example I can pass QuerySet as instance, while you strictly require List of dicts for some reason.

**Comment by kalekseev** [comment]
Serializers can be instantiated with instance = None, so it must be Optional, e.g.:
```
class Serializer(Generic[_IN]):
    def __init__(self, instance: Optional[_IN], ...):
```

File: rest_framework-stubs/serializers.pyi (hunk 4 above)
Comment (kalekseev) [reply rest_framework-stubs/serializers.pyi:168]: also why in init `instance: Union[MultiValueDict, Sequence[Any]] = ...,`?

File: rest_framework-stubs/serializers.pyi (hunk 1 above)
Comment (Goldziher) [reply rest_framework-stubs/serializers.pyi:215]: No time now...

File: rest_framework-stubs/serializers.pyi (hunk 2 above)
Comment (Goldziher) [reply rest_framework-stubs/serializers.pyi:143]: I disagree.

1. Model serializer is a specialist class for models.
2. We do not need to account for what users might do, but for the drf
3. We need proper type checking

File: rest_framework-stubs/serializers.pyi (hunk 3 above)
Comment (Goldziher) [reply rest_framework-stubs/serializers.pyi:91]: Its not used as often in field

File: rest_framework-stubs/serializers.pyi (hunk 4 above)
Comment (Goldziher) [reply rest_framework-stubs/serializers.pyi:168]: This is based on the DRF tests and in codes comments.

Comment (Goldziher) [reply rest_framework-stubs/serializers.pyi]: changed

**Comment by Goldziher** [comment]
Ok, all comments addressed in one way or another. I do not have time to work on this now for the remainder of the week. I would like to see this released though so I can actually use the stuff Ive worked on, so please consider proceeding @sobolevn

**Review by sobolevn — APPROVED** [review]

**Comment by sobolevn** [comment]
Let's have a bit of time for this to vent, before releasing it to the public. Since we don't practice `-rc` or anything similar.
You can now use https://github.com/typeddjango/djangorestframework-stubs/commit/c5dbc028662748cc0119477f766db2dc190c2246 as a reference to install it from `git`.

Thanks a lot for your hard work! 👍

**Pull request merged**